File: lib_base/src/main/java/com/android/base/utils/android/text/RoundedBackgroundSpan.java
package com.android.base.utils.android.text;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.text.style.ReplacementSpan;

import androidx.annotation.NonNull;

/**
 * 圆角背景Span
 */
public class RoundedBackgroundSpan extends ReplacementSpan {

    private int mPadding;
    private int mBackgroundColor;
    private int mTextColor;
    private int mCorner;

    public RoundedBackgroundSpan(int backgroundColor, int textColor, int padding, int corner) {
        super();
        mBackgroundColor = backgroundColor;
        mTextColor = textColor;
        mCorner = corner;
        mPadding = padding;
    }

    @Override
    public int getSize(@NonNull Paint paint, CharSequence text, int start, int end, Paint.FontMetricsInt fm) {
        return (int) (mPadding + paint.measureText(text.subSequence(start, end).toString()) + mPadding);
    }

    @Override
    public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom, @NonNull Paint paint) {
        float width = paint.measureText(text.subSequence(start, end).toString());
        RectF rect = new RectF(x, top, x + width + (2 * mPadding), bottom);
        paint.setColor(mBackgroundColor);
        canvas.drawRoundRect(rect, mCorner, mCorner, paint);
        paint.setColor(mTextColor);
        canvas.drawText(text, start, end, x + mPadding, y, paint);
    }

}

File: lib_base/src/main/java/com/android/base/utils/android/views/RoundedBackgroundSpan.kt
package com.android.base.utils.android.views

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.text.style.ReplacementSpan

/**
 * 圆角背景Span
 */
class RoundedBackgroundSpan(private val backgroundColor: Int, private val textColor: Int, private val padding: Int, private val corner: Int) : ReplacementSpan() {

    override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fm: Paint.FontMetricsInt?): Int {
        return (padding.toFloat() + paint.measureText(text.subSequence(start, end).toString()) + padding.toFloat()).toInt()
    }

    override fun draw(canvas: Canvas, text: CharSequence, start: Int, end: Int, x: Float, top: Int, y: Int, bottom: Int, paint: Paint) {
        val width = paint.measureText(text.subSequence(start, end).toString())
        val rect = RectF(x, top.toFloat(), x + width + (2 * padding).toFloat(), bottom.toFloat())
        paint.color = backgroundColor
        canvas.drawRoundRect(rect, corner.toFloat(), corner.toFloat(), paint)
        paint.color = textColor
        canvas.drawText(text, start, end, x + padding, y.toFloat(), paint)
    }

}